crates/datasources/src/json/table.rs

```diff
@@ -0,0 +1,135 @@
+use std::sync::Arc;
+use std::vec::Vec;
+
+use datafusion::arrow::datatypes::{DataType, Field, Schema};
+use datafusion::datasource::streaming::StreamingTable;
+use datafusion::datasource::TableProvider;
+use datafusion::physical_plan::streaming::PartitionStream;
+use serde_json::{Map, Value};
+
+use super::stream::JsonPartitionStream;
+use crate::common::url::DatasourceUrl;
+use crate::json::errors::JsonError;
+use crate::object_store::generic::GenericStoreAccess;
+use crate::object_store::ObjStoreAccess;
+
+pub async fn json_streaming_table(
+    store_access: GenericStoreAccess,
+    source_url: DatasourceUrl,
+) -> Result<Arc<dyn TableProvider>, JsonError> {
+    let path = source_url.path();
+
+    let store = store_access.create_store()?;
+
+    // assume that the file type is a glob and see if there are
+    // more files...
+    let mut list = store_access.list_globbed(&store, path.as_ref()).await?;
+
+    if list.is_empty() {
+        return Err(JsonError::NotFound(path.into_owned()));
+    }
+
+    // for consistent results, particularly for the sample, always
+    // sort by location
+    list.sort_by(|a, b| a.location.cmp(&b.location));
+
+    let mut data = Vec::new();
+    for obj in list {
+        let blob = store.get(&obj.location).await?.bytes().await?.to_vec();
+        let dejson = serde_json::from_slice::<serde_json::Value>(blob.as_slice())?.to_owned();
+        push_unwind_json_values(&mut data, dejson)?;
+    }
+
+    let mut field_set = indexmap::IndexMap::<String, DataType>::new();
+    for obj in &data {
+        for key in obj.keys() {
+            if field_set.contains_key(key) {
+                continue;
+            }
+            field_set.insert(key.to_string(), type_for_value(obj.get(key).unwrap()));
+        }
+    }
+
+    let schema = Arc::new(Schema::new(
+        field_set
+            .iter()
+            .map(|(k, v)| Field::new(k, v.to_owned(), true))
+            .collect::<Vec<_>>(),
+    ));
+
+    let chunks = data
+        .chunks(100)
```

Comment:
100 chunks seems like a pretty small default.

Reply:
isn't this "chunks of 100" (likely I think there aren't likely to be relatively few documents in any of these tables)

Reply:
how big should chunks be? 

I think in general, being able to split this into a number of partitions is good, and means processing could be meaningfully async, but it'd be reasonable to just have one too.

Reply:
when working on the polars json parser, we ended up landing on `1000` as the default chunk size. I don't have the benchmarks anymore, but it seemed to be the sweet spot in partition sizes for most cases.

Reply:
> I think in general, being able to split this into a number of partitions is good, and means processing could be meaningfully async, but it'd be reasonable to just have one too.

Don't the chunks already contain the deserialized data, which is the most expensive part. Isn't the stream just creating recordbatches from the chunks? If we were to defer the cpu bound tasks (deserialization) to the stream, then I think this would make more sense to chunk. In this scenario, I think the chunking may introduce more partitioning overhead than benefits. _(we'd probably want to benchmark it to be sure, but that's just my suspicion)._

Reply:
I'm down for 1000